File: modules/checkpoint.py
"""Crash/ban-safe progress tracking for multi-target runs.

A long unattended scan of many bank domains can die mid-way — antifraud ban,
expired session, killed shell. The checkpoint records which targets finished so
`--resume` can pick up the rest instead of re-scanning everything (and re-poking
antifraud on domains already covered).

State lives in `<output_dir>/.vectrix_state/<mode>.json` and is rewritten after
every completed target, so at most one target's progress is ever lost.
"""
from __future__ import annotations

import json
from pathlib import Path


class Checkpoint:
    def __init__(self, output_dir: Path, mode: str):
        self.dir = output_dir / ".vectrix_state"
        self.path = self.dir / f"{mode}.json"
        self.mode = mode
        self._completed: set[str] = set()

    # ── Lifecycle ────────────────────────────────────────────────────────────

    def load(self) -> None:
        """Load previously completed targets (for --resume)."""
        if not self.path.exists():
            return
        try:
            data = json.loads(self.path.read_text(encoding="utf-8"))
            self._completed = set(data.get("completed", []))
        except (json.JSONDecodeError, OSError):
            self._completed = set()

    def reset(self, targets: list[str]) -> None:
        """Begin a fresh run (no resume): clear prior completion state."""
        self._completed = set()
        self._write(targets)

    def is_done(self, target: str) -> bool:
        return target in self._completed

    def pending(self, targets: list[str]) -> list[str]:
        return [t for t in targets if t not in self._completed]

    def mark_done(self, target: str, targets: list[str]) -> None:
        self._completed.add(target)
        self._write(targets)

    def clear(self) -> None:
        """Remove the state file once the whole run finished cleanly."""
        try:
            self.path.unlink()
        except OSError:
            pass

    # ── Internal ─────────────────────────────────────────────────────────────

    def _write(self, targets: list[str]) -> None:
        self.dir.mkdir(parents=True, exist_ok=True)
        payload = {
            "mode": self.mode,
            "targets": targets,
            "completed": sorted(self._completed),
        }
        tmp = self.path.with_suffix(".json.tmp")
        tmp.write_text(json.dumps(payload, indent=2, ensure_ascii=False), encoding="utf-8")
        tmp.replace(self.path)

File: go-vectrix/internal/checkpoint/checkpoint.go
// Package checkpoint provides crash/ban-safe progress tracking for
// multi-target runs, mirroring modules/checkpoint.py.
//
// State lives in <output_dir>/.vectrix_state/<mode>.json and is rewritten
// after every completed target, so at most one target's progress is lost if
// the run dies mid-way.
package checkpoint

import (
	"encoding/json"
	"os"
	"path/filepath"
	"sort"
)

type Checkpoint struct {
	dir       string
	path      string
	mode      string
	completed map[string]struct{}
}

type fileFormat struct {
	Mode      string   `json:"mode"`
	Targets   []string `json:"targets"`
	Completed []string `json:"completed"`
}

// New creates a checkpoint tracker for outputDir/.vectrix_state/<mode>.json.
func New(outputDir, mode string) *Checkpoint {
	dir := filepath.Join(outputDir, ".vectrix_state")
	return &Checkpoint{
		dir:       dir,
		path:      filepath.Join(dir, mode+".json"),
		mode:      mode,
		completed: make(map[string]struct{}),
	}
}

// Load reads previously completed targets (for --resume).
func (c *Checkpoint) Load() {
	data, err := os.ReadFile(c.path)
	if err != nil {
		return
	}
	var f fileFormat
	if err := json.Unmarshal(data, &f); err != nil {
		return
	}
	c.completed = make(map[string]struct{}, len(f.Completed))
	for _, t := range f.Completed {
		c.completed[t] = struct{}{}
	}
}

// Reset begins a fresh run: clears prior completion state and writes targets.
func (c *Checkpoint) Reset(targets []string) {
	c.completed = make(map[string]struct{})
	c.write(targets)
}

// IsDone reports whether target was already completed.
func (c *Checkpoint) IsDone(target string) bool {
	_, ok := c.completed[target]
	return ok
}

// Pending returns the subset of targets not yet completed.
func (c *Checkpoint) Pending(targets []string) []string {
	out := make([]string, 0, len(targets))
	for _, t := range targets {
		if !c.IsDone(t) {
			out = append(out, t)
		}
	}
	return out
}

// MarkDone records target as completed and persists state.
func (c *Checkpoint) MarkDone(target string, targets []string) {
	c.completed[target] = struct{}{}
	c.write(targets)
}

// Clear removes the state file once the whole run finished cleanly.
func (c *Checkpoint) Clear() {
	os.Remove(c.path)
}

func (c *Checkpoint) write(targets []string) {
	if err := os.MkdirAll(c.dir, 0o755); err != nil {
		return
	}
	completed := make([]string, 0, len(c.completed))
	for t := range c.completed {
		completed = append(completed, t)
	}
	sort.Strings(completed)

	data, err := json.MarshalIndent(fileFormat{
		Mode:      c.mode,
		Targets:   targets,
		Completed: completed,
	}, "", "  ")
	if err != nil {
		return
	}
	tmp := c.path + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return
	}
	os.Rename(tmp, c.path)
}
